File: scripts/lib/platform-paths.mjs
/**
 * JustSearch platform-paths — Node implementation of the
 * `contracts/platform-paths/spec.v1.json` contract (tempdoc 501 §3.6).
 *
 * This module is the single Node source of truth for resolving the
 * JustSearch data directory. It replaces the drifting copies that
 * previously lived in `modules/ui-web/vite.config.js`,
 * `scripts/dev/dev-runner.cjs`, and `scripts/prod/justsearch-mcp/discovery.mjs`.
 *
 * The Java reference is `modules/configuration/.../PlatformPaths.java`.
 * The cross-language `PlatformPathsContractTest` feeds the same fixture set
 * into every implementation and asserts identical outputs.
 *
 * Node has no JVM system properties — sysprop-class sources in the spec are
 * skipped here. The env-var + platform-default branches are authoritative
 * for the Node side.
 */

import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const SPEC_PATH = path.resolve(__dirname, '..', '..', 'contracts', 'platform-paths', 'spec.v1.json');

/** Lazily-loaded contract spec (for callers wanting to introspect or test). */
let _spec = null;
export function loadSpec() {
  if (_spec === null) {
    _spec = JSON.parse(fs.readFileSync(SPEC_PATH, 'utf8'));
  }
  return _spec;
}

/**
 * Replace literal `${user.home}` placeholders. Returns the input unchanged if
 * no placeholder is present. Refuses to leave unexpanded placeholders.
 */
export function expandUserHomePlaceholders(raw, userHomeOverride = null) {
  if (raw == null || raw === '') return raw;
  if (!raw.includes('${user.home}')) return raw;
  const userHome = userHomeOverride ?? os.homedir();
  if (!userHome) return raw;
  return raw.replace(/\$\{user\.home\}/g, userHome);
}

/** Detect the canonical platform key used by the contract spec. */
export function detectPlatform(platformOverride = null) {
  const p = platformOverride ?? process.platform;
  // Already a contract-spec name (passed from a fixture/test) — return as-is.
  if (p === 'windows' || p === 'macos' || p === 'linux') return p;
  // node process.platform aliases.
  if (p === 'win32') return 'windows';
  if (p === 'darwin') return 'macos';
  return 'linux';
}

/**
 * Resolve the data dir following `contracts/platform-paths/spec.v1.json`.
 *
 * Accepts an optional override bundle for testing — pass {env, userHome,
 * platform} to feed the contract fixtures into this function without
 * touching the actual process environment.
 */
export function resolveDataDir({env = process.env, userHome = null, platform = null} = {}) {
  const home = userHome ?? os.homedir();
  const plat = detectPlatform(platform);

  // 1. Env-var sources (in spec order). sysprop sources are JVM-only.
  const envValue = env['JUSTSEARCH_DATA_DIR'];
  if (envValue && envValue.trim()) {
    return expandUserHomePlaceholders(envValue.trim(), home);
  }

  // 2. Platform default.
  if (plat === 'windows') {
    const localAppData = env['LOCALAPPDATA'];
    if (localAppData && localAppData.trim()) {
      return path.join(localAppData, 'JustSearch');
    }
    return path.join(home, 'AppData', 'Local', 'JustSearch');
  }
  if (plat === 'macos') {
    return path.join(home, 'Library', 'Application Support', 'JustSearch');
  }
  return path.join(home, '.justsearch');
}

/** Convenience: data dir resolved against the live process environment. */
export function resolveDataDirLive() {
  return resolveDataDir();
}

/** Convenience for the manifest readers: `<dataDir>/runtime/manifest.json`. */
export function resolveManifestPath(dataDir = null) {
  const d = dataDir ?? resolveDataDirLive();
  return path.join(d, 'runtime', 'manifest.json');
}

/**
 * Convenience: read the runtime manifest from the resolved data directory,
 * returning the parsed JSON or null if the file is absent / malformed.
 */
export function readManifestSync(dataDir = null) {
  try {
    const content = fs.readFileSync(resolveManifestPath(dataDir), 'utf8');
    return JSON.parse(content);
  } catch {
    return null;
  }
}

/**
 * Check if a PID is alive on the current OS. Returns true if alive, false if
 * dead, null if we can't tell (lacking permissions etc.).
 *
 * Implementation: process.kill(pid, 0) throws ESRCH for dead, EPERM for alive
 * without permission, no-throw for alive-with-permission. We treat the no-perm
 * case as alive (safer assumption: don't claim a stale manifest is fresh).
 */
export function isPidAlive(pid) {
  if (typeof pid !== 'number' || !Number.isFinite(pid) || pid <= 0) return false;
  try {
    process.kill(pid, 0);
    return true;
  } catch (e) {
    if (e.code === 'ESRCH') return false;
    if (e.code === 'EPERM') return true;
    return null;
  }
}

/**
 * Find a running JustSearch backend by walking a prioritized list of
 * candidate data directories, reading each one's runtime manifest, and
 * returning the first manifest whose `pid` is alive on this OS.
 *
 * Why a candidate list rather than the single resolveDataDir result: under
 * bare `npm run dev` (no orchestrator-injected JUSTSEARCH_DATA_DIR) the
 * canonical resolver falls through to the platform default, which is
 * usually a stale production install. Dev backends bind to a worktree-local
 * `.dev-data` instead. The resolver's algorithm is correct; the candidate
 * list extends it with dev conventions documented in this function.
 *
 * @param {object} [opts]
 * @param {string} [opts.cwd] — starting point for worktree/repo-root probing
 * @param {string[]} [opts.extraDataDirs] — caller-supplied data dirs probed
 *   FIRST (e.g., a Vite worktree's `modules/ui-web/.dev-data`).
 * @returns {{manifest: object, dataDir: string} | null}
 */
export function findRunningManifest({cwd = process.cwd(), extraDataDirs = []} = {}) {
  const candidates = [];

  for (const d of extraDataDirs) {
    if (d) candidates.push(d);
  }

  if (process.env.JUSTSEARCH_DATA_DIR && process.env.JUSTSEARCH_DATA_DIR.trim()) {
    candidates.push(process.env.JUSTSEARCH_DATA_DIR.trim());
  }

  // Dev convention: walk up from cwd looking for `modules/ui-web/.dev-data` or
  // `.dev-data` at the repo root.
  let walk = path.resolve(cwd);
  for (let i = 0; i < 8; i++) {
    candidates.push(path.join(walk, 'modules', 'ui-web', '.dev-data'));
    candidates.push(path.join(walk, '.dev-data'));
    const parent = path.dirname(walk);
    if (parent === walk) break;
    walk = parent;
  }

  // Platform default (covers production installs).
  candidates.push(resolveDataDir());

  const tried = [];
  for (const dataDir of candidates) {
    const manifest = readManifestSync(dataDir);
    if (!manifest) {
      tried.push({dataDir, status: 'no-manifest'});
      continue;
    }
    const alive = isPidAlive(manifest.pid);
    if (alive === false) {
      tried.push({dataDir, status: 'stale-pid', pid: manifest.pid});
      continue;
    }
    return {manifest, dataDir};
  }
  return null;
}

File: modules/shell/src-tauri/src/platform_paths.rs
//! Rust implementation of the `contracts/platform-paths/spec.v1.json` contract
//! (tempdoc 501 §3.6).
//!
//! This module is the single Rust source of truth for resolving the JustSearch
//! data directory. Mirrors the precedence + platform-default logic from the
//! reference Java implementation
//! (`modules/configuration/.../PlatformPaths.java`) and the Node twin
//! (`scripts/lib/platform-paths.mjs`).
//!
//! The shell does not run inside a JVM, so sysprop-class sources from the
//! contract are skipped here (same convention as the Node twin). Env-var +
//! platform-default branches are authoritative.

use std::env;
use std::path::PathBuf;

/// Replace literal `${user.home}` placeholders. Returns the input unchanged if
/// no placeholder is present. Mirrors PlatformPaths.expandUserHomePlaceholders.
pub fn expand_user_home_placeholders(raw: &str) -> String {
    if !raw.contains("${user.home}") {
        return raw.to_string();
    }
    if let Some(home) = user_home() {
        return raw.replace("${user.home}", &home);
    }
    raw.to_string()
}

fn user_home() -> Option<String> {
    // dirs-style helper without bringing in the dirs crate. On Windows
    // USERPROFILE / HOMEDRIVE+HOMEPATH; elsewhere HOME.
    if cfg!(target_os = "windows") {
        if let Ok(p) = env::var("USERPROFILE") {
            if !p.is_empty() {
                return Some(p);
            }
        }
        let drive = env::var("HOMEDRIVE").ok();
        let dir = env::var("HOMEPATH").ok();
        if let (Some(d), Some(h)) = (drive, dir) {
            if !d.is_empty() && !h.is_empty() {
                return Some(format!("{d}{h}"));
            }
        }
    }
    env::var("HOME").ok().filter(|s| !s.is_empty())
}

/// Resolve the data directory per the shared contract.
///
/// Precedence:
///   1. `JUSTSEARCH_DATA_DIR` env var (expands `${user.home}` if present)
///   2. Platform default:
///        windows: `%LOCALAPPDATA%/JustSearch` (or `userHome/AppData/Local/JustSearch`)
///        macos:   `userHome/Library/Application Support/JustSearch`
///        linux:   `userHome/.justsearch`
pub fn resolve_data_dir() -> PathBuf {
    if let Ok(env_val) = env::var("JUSTSEARCH_DATA_DIR") {
        let trimmed = env_val.trim();
        if !trimmed.is_empty() {
            return PathBuf::from(expand_user_home_placeholders(trimmed));
        }
    }
    platform_default()
}

fn platform_default() -> PathBuf {
    if cfg!(target_os = "windows") {
        if let Ok(la) = env::var("LOCALAPPDATA") {
            if !la.is_empty() {
                return PathBuf::from(la).join("JustSearch");
            }
        }
        let home = user_home().unwrap_or_else(|| "C:\\".to_string());
        return PathBuf::from(home)
            .join("AppData")
            .join("Local")
            .join("JustSearch");
    }
    if cfg!(target_os = "macos") {
        let home = user_home().unwrap_or_else(|| "/".to_string());
        return PathBuf::from(home)
            .join("Library")
            .join("Application Support")
            .join("JustSearch");
    }
    let home = user_home().unwrap_or_else(|| "/".to_string());
    PathBuf::from(home).join(".justsearch")
}

/// Filesystem path the producer writes its manifest to.
pub fn resolve_manifest_path() -> PathBuf {
    resolve_data_dir().join("runtime").join("manifest.json")
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::collections::HashMap;
    use std::fs;
    use std::path::PathBuf;
    use std::sync::Mutex;

    // Tempdoc 501 §3.6: contract-spec fixtures mutate process-global env vars
    // hermetically. Cargo runs tests in parallel by default; serialize all
    // env-touching tests through one mutex so they don't observe each other's
    // overrides. The lock is the only state shared across these tests.
    static ENV_LOCK: Mutex<()> = Mutex::new(());

    #[test]
    fn expand_user_home_returns_input_when_no_placeholder() {
        let _g = ENV_LOCK.lock().unwrap();
        assert_eq!(expand_user_home_placeholders("/var/lib"), "/var/lib");
    }

    /// Tempdoc 501 §3.6: drive the cross-language contract spec
    /// (`contracts/platform-paths/spec.v1.json`) through the Rust resolver.
    /// Fixtures marked `runners: [...]` that omit "rust" are skipped, mirroring
    /// the Java + Node twin tests.
    #[test]
    fn resolve_data_dir_matches_contract_spec() {
        let _g = ENV_LOCK.lock().unwrap();
        let spec_path = find_spec_path().expect("spec.v1.json not found");
        let spec_text = fs::read_to_string(&spec_path).expect("read spec");
        let spec: serde_json::Value = serde_json::from_str(&spec_text).expect("parse spec");
        let fixtures = spec
            .get("fixtures")
            .and_then(|f| f.as_array())
            .expect("fixtures array");

        let mut ran = 0;
        let mut failures: Vec<String> = Vec::new();
        for fixture in fixtures {
            let name = fixture
                .get("name")
                .and_then(|v| v.as_str())
                .unwrap_or("<unnamed>");

            // Filter by runners marker; rust-only/missing-marker fixtures run.
            if let Some(runners) = fixture.get("runners").and_then(|r| r.as_array()) {
                if !runners.iter().any(|r| r.as_str() == Some("rust")) {
                    continue;
                }
            }

            // Skip fixtures targeting platforms other than the current host —
            // resolve_data_dir branches on cfg!(target_os), which can't be
            // overridden at runtime. Linux-fixtures run on Linux hosts; etc.
            let platform = fixture.get("platform").and_then(|v| v.as_str()).unwrap_or("");
            if !matches_current_platform(platform) {
                continue;
            }

            // Hermetic env override under the global mutex.
            let env = fixture
                .get("env")
                .and_then(|v| v.as_object())
                .map(|m| {
                    m.iter()
                        .map(|(k, v)| (k.clone(), v.as_str().unwrap_or("").to_string()))
                        .collect::<HashMap<_, _>>()
                })
                .unwrap_or_default();
            let user_home = fixture
                .get("userHome")
                .and_then(|v| v.as_str())
                .unwrap_or("");
            let expected = fixture
                .get("expected")
                .and_then(|v| v.as_str())
                .unwrap_or("");

            let actual = with_env_overrides(&env, user_home, || resolve_data_dir());
            let actual_norm = actual.to_string_lossy().replace('\\', "/");
            let expected_norm = expected.replace('\\', "/");

            if actual_norm != expected_norm {
                failures.push(format!(
                    "{name}: expected {expected_norm}, got {actual_norm}"
                ));
            }
            ran += 1;
        }

        assert!(
            failures.is_empty(),
            "{} of {} rust fixtures failed:\n{}",
            failures.len(),
            ran,
            failures.join("\n"),
        );
        assert!(ran > 0, "no rust-targeted fixtures executed — harness regression");
    }

    fn matches_current_platform(p: &str) -> bool {
        match p {
            "linux" => cfg!(target_os = "linux"),
            "macos" => cfg!(target_os = "macos"),
            "windows" => cfg!(target_os = "windows"),
            "" => true,
            _ => false,
        }
    }

    /// Save+set+restore the affected env vars and HOME under a closure.
    /// Always runs under `ENV_LOCK` so cargo's parallel test runner can't
    /// race two fixtures' env-touching scopes.
    fn with_env_overrides<F, R>(
        env: &HashMap<String, String>,
        user_home: &str,
        f: F,
    ) -> R
    where
        F: FnOnce() -> R,
    {
        let touched_keys: Vec<&str> = env
            .keys()
            .map(|s| s.as_str())
            .chain(std::iter::once("HOME"))
            .chain(std::iter::once("USERPROFILE"))
            .chain(std::iter::once("LOCALAPPDATA"))
            .collect::<std::collections::BTreeSet<_>>()
            .into_iter()
            .collect();

        let saved: HashMap<&str, Option<String>> = touched_keys
            .iter()
            .map(|k| (*k, std::env::var(*k).ok()))
            .collect();

        for k in &touched_keys {
            std::env::remove_var(*k);
        }
        if !user_home.is_empty() {
            std::env::set_var("HOME", user_home);
            // Windows-style fallback chain too — resolve_data_dir checks
            // USERPROFILE first.
            std::env::set_var("USERPROFILE", user_home);
        }
        for (k, v) in env {
            std::env::set_var(k, v);
        }

        let result = f();

        // Restore.
        for k in &touched_keys {
            match saved.get(*k).and_then(|v| v.clone()) {
                Some(v) => std::env::set_var(*k, v),
                None => std::env::remove_var(*k),
            }
        }
        result
    }

    fn find_spec_path() -> Option<PathBuf> {
        let mut cur = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
        for _ in 0..8 {
            let candidate = cur.join("contracts/platform-paths/spec.v1.json");
            if candidate.is_file() {
                return Some(candidate);
            }
            let parent = cur.parent()?.to_path_buf();
            cur = parent;
        }
        None
    }
}
